nncf/quantization/passes.py

```diff
@@ -0,0 +1,69 @@
+"""
+ Copyright (c) 2023 Intel Corporation
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+import collections
+
+from nncf.common.graph.graph import NNCFGraph
+from nncf.common.graph.operator_metatypes import ShapeOfMetatype
+
+
+def transform_to_inference_graph(nncf_graph: NNCFGraph) -> NNCFGraph:
+    """
+    This method contains pipeline of the passes that uses to provide inference graph without constant flows.
+
+    :param nncf_graph: NNCFGraph instance for the transformation.
+    :return: NNCFGraph in the inference style.
+    """
+    inference_nncf_graph = remove_shape_of_subgraphs(nncf_graph)
+    return inference_nncf_graph
+
+
+def remove_shape_of_subgraphs(nncf_graph: NNCFGraph) -> NNCFGraph:
+    """
+    Removes the ShapeOf subgraphs from the provided NNCFGraph instance.
+
+    :param nncf_graph: NNCFGraph instance for the transformation.
+    :return: NNCFGraph without ShapeOf subgraphs.
+    """
+    nodes_to_drop = set()
+    shape_of_nodes = []
+    infer_nodes = []
+
+    nodes_queue = collections.deque()
+    nodes_queue.extend(nncf_graph.get_input_nodes())
+    while nodes_queue:
+        node = nodes_queue.pop()
+        if issubclass(node.metatype, ShapeOfMetatype):
+            shape_of_nodes.append(node)
+            continue
+        if node in infer_nodes:
+            continue
+        infer_nodes.append(node)
+        nodes_queue.extend(nncf_graph.get_next_nodes(node))
+
+    for shape_of_node in shape_of_nodes:
+        nodes_to_drop.add(shape_of_node)
+
+        shape_of_queue = collections.deque()
+        shape_of_queue.extend(nncf_graph.get_next_nodes(shape_of_node))
+        while shape_of_queue:
+            node = shape_of_queue.pop()
+            if node in nodes_to_drop or node in infer_nodes:
```

Comment:
`node in infer_nodes` -- `O(len(infer_nodes))`

Is it possible to use `dict` instead of a `list` to check this?

Reply:
It's possible. But I don't see any significant performance improvements from this.

Reply:
ok